ts/tree/dbTree.ts

```diff
@@ -1,17 +1,218 @@
-import { Hashable } from "../interfaces";
-import { Hasher } from "./hasher";
-import { Leaf, LeafFactoryFunc } from "./leaves/Leaf";
-import { MemoryTree } from "./memoryTree";
-
-export class DBTree<LeafType extends Leaf<Hashable>> extends MemoryTree {
-    private readonly leafFactory: LeafFactoryFunc<LeafType>;
-
-    constructor(
-        depth: number,
-        leafFactory: LeafFactoryFunc<LeafType>,
-        hasher?: Hasher
-    ) {
-        super(depth, hasher || Hasher.new());
-        this.leafFactory = leafFactory;
+import {
+    BadMergeAlignment,
+    EmptyArray,
+    ExceedTreeSize,
+    MismatchHash,
+    MismatchLength,
+    NegativeIndex
+} from "../exceptions";
+import { Hasher, Node } from "./hasher";
+import { ItemNode } from "./leaves/Node";
+import { AsyncTree, Data, Witness } from "./tree";
+
+export class DBTree implements AsyncTree {
+    public readonly zeros: Array<Node>;
+    public readonly depth: number;
+    public readonly setSize: number;
+    public readonly hasher: Hasher;
+    public readonly nodeType: string;
+    public _root: Node;
+
+    public static new(depth: number, nodeType: string, hasher?: Hasher) {
+        return new DBTree(depth, nodeType, hasher || Hasher.new());
+    }
+
+    constructor(depth: number, nodeType: string, hasher: Hasher) {
+        this.depth = depth;
+        this.hasher = hasher;
+        this.zeros = this.hasher.zeros(depth);
+        this.setSize = 2 ** this.depth;
+        this.nodeType = nodeType;
+        this._root = this.zeros[0];
```

Comment:
I just comment to highlight issues here. We can keep it as is in this PR and fix the problem in the next PR.

This works when we have an empty tree. But we will have a problem once we have new leaves in the tree and have a non-empty tree root in DB.

When we restart the client and read the `root()` from the DBTree, we expect to read the non-empty tree root in DB. But here the cached root still returns the empty tree root.

We will need 2 steps. One reads the cache from DB and a step dumps the cache to DB.

Reply:
absolutely! I had planned on discussing this on the call. We can come up with a `restoreFromDB` function which can check if a tree already exists and if yes, it can load the tree root. Was a bit confused though as to where we will be calling that.